bigtable-hbase-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/read/FlatRowAdapter.java

```diff
@@ -88,79 +78,87 @@ private Result adaptResponse(FlatRow flatRow, final boolean sort) {
    * {@link WhileMatchFilter}
    * <p>
    * Some additional deduping and sorting is also required. Cloud Bigtable returns values sorted by
-   * family in lexicographically ascending order, an arbitrary ordering of qualifiers (by internal
-   * id rather than lexicographically) and finally by timestamp descending. Each cell can appear
-   * more than once, if there are {@link Interleave}s in the {@link ReadRowsRequest#getFilter()},
-   * but the duplicates will appear one after the other. An {@link Interleave}s can originate from a
-   * {@link FilterList} with a {@link FilterList.Operator#MUST_PASS_ONE} operator or a
-   * {@link WhileMatchFilter}.
+   * family (by internal id, not lexicographically), a lexicographically ascending ordering of
+   * qualifiers, and finally by timestamp descending. Each cell can appear more than once, if there
+   * are {@link Interleave}s in the {@link ReadRowsRequest#getFilter()}, but the duplicates will
+   * appear one after the other. An {@link Interleave}s can originate from a {@link FilterList} with
+   * a {@link FilterList.Operator#MUST_PASS_ONE} operator or a {@link WhileMatchFilter}.
    * <p>
    * @param flatRow
-   * @param sort
-   * @return
+   * @return a
    */
-  private List<Cell> extractCells(FlatRow flatRow, boolean sort) {
+  private Cell[] extractCellsSortAndDedup(FlatRow flatRow) {
     byte[] rowKey = ByteStringer.extract(flatRow.getRowKey());
 
     FlatRow.Cell previousCell = null;
     byte[] previousFamily = null;
 
-    List<FlatRow.Cell> cells = flatRow.getCells();
-    List<Cell> hbaseCells = new ArrayList<>(cells.size());
-    List<RowCell> familyCells = sort ? new ArrayList<RowCell>(cells.size()) : null;
+    Map<String, List<RowCell>> familyMap = new TreeMap<>();
+    List<RowCell> currentFamilyRowCells = new ArrayList<>(flatRow.getCells().size());
```

Comment:
Your point about readability is well taken.  There's too much going on in that method.  Let me see what I can do to improve readability while maintaining performance.